readme.md
This is a simple perl binding for connecting to [Atomia Automation Server](http://www.atomia.com) CoreAPI.

It supports all methods, and includes helper classes for easing the SOAP serialization pain of parameters with
one of the following data types:

* ProvisioningService
* ServiceSearchCriteria[]

This will be extended in the future as needed, if you need some other data type, feel free to add and
send merge request.

## Example

```perl
        my $coreapi = Atomia::Bindings::CoreAPI->new(
                endpoint => "https://some.provisioning.host/CoreAPIBasicAuth.svc",
                username => "Administrator", password => "somepass");

        my $services = $coreapi->FindServicesByPath(
                Atomia::Bindings::CoreAPI::ServiceSearchCriteriaArray->new(criterias => [ {
                        parent_service => undef,
                        service_name => "CsInstances",
                        service_path => "CsCloud",
                } ]),
                Atomia::Bindings::CoreAPI::Account->new(name => 100052));

        my $cs_instances = undef;
        if (defined($services) && ref($services) eq "HASH") {
                $services = $services->{"ProvisioningService"};
                $services = [ $services ] if defined($services) && ref($services) ne 'ARRAY';
                $cs_instances = Atomia::Bindings::CoreAPI::ProvisioningService->new(
                        element_name => "parentService",
                        fetched => $services->[0]) if defined($services) && scalar($services) > 0;
        }

        die "CsInstances not found" unless defined($cs_instances);

        my $new_service = $coreapi->CreateService(Atomia::Bindings::CoreAPI::ServiceName->new(
                name => "CsLinuxInstance"),
                $cs_instances,
                Atomia::Bindings::CoreAPI::Account->new(element_name => "accountName", name => 100052));

        $new_service = Atomia::Bindings::CoreAPI::ProvisioningService->new(fetched => $new_service);
        $new_service->setprop("Name", "autoscaletest_node1");
        $new_service->setprop("InstanceType", "m1.tiny");
        $new_service->setprop("ImageId", "ami-00000002");
        $new_service->setprop("KeyPair", "democluster");
        $new_service->setprop("GlobalSecurityGroup", "webservers");
        $new_service->setprop("AgentPassword", "foo");
        $coreapi->AddService($new_service, $cs_instances,
                Atomia::Bindings::CoreAPI::Account->new(element_name => "accountName", name => 100052));

        print "New instance provisioned successfully\n";
```


debian/changelog
libatomia-bindings-coreapi-perl (1.0.3) lucid; urgency=low

  * Make soap timeout configurable and increase default

 -- Jimmy Bergman <jimmy@sigint.se>  Thu, 4 Aug 2011 16:20:21 +0200

libatomia-bindings-coreapi-perl (1.0.2) lucid; urgency=low

  * Fix ServiceSearchCriteria serialization when parent_service is set

 -- Jimmy Bergman <jimmy@sigint.se>  Tue, 2 Aug 2011 09:50:04 +0200

libatomia-bindings-coreapi-perl (1.0.1) lucid; urgency=low

  * Add getprops

 -- Jimmy Bergman <jimmy@sigint.se>  Mon, 1 Aug 2011 16:13:09 +0200

libatomia-bindings-coreapi-perl (1.0.0) lucid; urgency=low

  * Initial version.

 -- Jimmy Bergman <jimmy@sigint.se>  Fri, 29 Jul 2011 14:40:13 +0200


examples/example.pl
#!/usr/bin/perl -w

use strict;
use warnings;

use Atomia::Bindings::CoreAPI;
use Data::Dumper;

die "usage: $0 adminpass" unless scalar(@ARGV) > 0;

my $coreapi = Atomia::Bindings::CoreAPI->new(endpoint => "https://provisioning.int.atomia.com/CoreAPIBasicAuth.svc", username => "Administrator", password => $ARGV[0], trace => 0);

eval {
	my $services = $coreapi->FindServicesByPath(
		Atomia::Bindings::CoreAPI::ServiceSearchCriteriaArray->new(criterias => [ {
			parent_service => undef,
			service_name => "CsLinuxInstance",
			service_path => "CsCloud/CsInstances",
		} ]),
		Atomia::Bindings::CoreAPI::Account->new(name => 100052));

	if (defined($services) && ref($services) eq "HASH") {
		$services = $services->{"ProvisioningService"};
		$services = [ $services ] if defined($services) && ref($services) ne 'ARRAY';

		if (defined($services)) {
			print scalar(@$services) . " services found, trying to delete the first\n";
			$coreapi->DeleteService(Atomia::Bindings::CoreAPI::ProvisioningService->new(fetched => $services->[0]), Atomia::Bindings::CoreAPI::Account->new(element_name => "accountName", name => 100052));
		}
	}

	$services = $coreapi->FindServicesByPath(
		Atomia::Bindings::CoreAPI::ServiceSearchCriteriaArray->new(criterias => [ {
			parent_service => undef,
			service_name => "CsInstances",
			service_path => "CsCloud",
		} ]),
		Atomia::Bindings::CoreAPI::Account->new(name => 100052));

	my $cs_instances = undef;
	if (defined($services) && ref($services) eq "HASH") {
		$services = $services->{"ProvisioningService"};
		$services = [ $services ] if defined($services) && ref($services) ne 'ARRAY';
		$cs_instances = Atomia::Bindings::CoreAPI::ProvisioningService->new(element_name => "parentService", fetched => $services->[0]) if defined($services) && scalar($services) > 0;
	}

	die "CsInstances not found" unless defined($cs_instances);

	my $new_service = $coreapi->CreateService(Atomia::Bindings::CoreAPI::ServiceName->new(name => "CsLinuxInstance"), $cs_instances, Atomia::Bindings::CoreAPI::Account->new(element_name => "accountName", name => 100052));
	$new_service = Atomia::Bindings::CoreAPI::ProvisioningService->new(fetched => $new_service);
	$new_service->setprop("Name", "autoscaletest_node1");
	$new_service->setprop("InstanceType", "m1.tiny");
	$new_service->setprop("ImageId", "ami-00000002");
	$new_service->setprop("KeyPair", "democluster");
	$new_service->setprop("GlobalSecurityGroup", "webservers");
	$new_service->setprop("AgentPassword", "foo");
	$coreapi->AddService($new_service, $cs_instances, Atomia::Bindings::CoreAPI::Account->new(element_name => "accountName", name => 100052));

	print "New instance provisioned successfully\n";
};

if ($@) {
	print "exception: " . $@->message . "\n";
}
